identity-actor/src/actor/actor_builder.rs

```diff
@@ -0,0 +1,274 @@
+      let noise_keypair = NoiseKeypair::<X25519Spec>::new().into_authentic(&keypair).unwrap();
+      let peer_id = keypair.public().to_peer_id();
+      (noise_keypair, peer_id)
+    };
+
+    let executor = Box::new(|fut| {
+      tokio::spawn(fut);
+    });
+
+    let mut swarm: Swarm<RequestResponse<ActorRequestResponseCodec>> = {
+      let mut config: RequestResponseConfig = RequestResponseConfig::default();
+      config.set_request_timeout(self.config.timeout);
+
+      let behaviour = RequestResponse::new(
+        ActorRequestResponseCodec(),
+        iter::once((ActorProtocol(), ProtocolSupport::Full)),
+        config,
+      );
+
+      let transport = transport
+        .upgrade(upgrade::Version::V1)
+        .authenticate(NoiseConfig::xx(noise_keypair).into_authenticated())
+        .multiplex(YamuxConfig::default())
+        .boxed();
+
+      SwarmBuilder::new(transport, behaviour, peer_id)
+        .executor(executor.clone())
+        .build()
+    };
+
+    for addr in self.listening_addresses {
+      swarm.listen_on(addr).map_err(|err| Error::TransportError {
+        context: "unable to start listening",
+        source: err,
+      })?;
+    }
+
+    let (cmd_sender, cmd_receiver) = mpsc::channel(10);
+
+    let event_loop = EventLoop::new(swarm, cmd_receiver);
+    let swarm_commander = NetCommander::new(cmd_sender);
+
+    let actor = Actor::from_builder(
+      swarm_commander.clone(),
+      self.handler_map,
+      self.object_map,
+      peer_id,
+      self.config,
+    )
+    .await?;
+
+    let actor_clone = actor.clone();
+
+    let event_handler = move |event: InboundRequest| {
+      let actor = actor_clone.clone();
+
+      if event.request_mode == RequestMode::Asynchronous {
+        actor.handle_request::<AsynchronousInvocationStrategy>(event);
+      } else {
+        actor.handle_request::<SynchronousInvocationStrategy>(event);
+      };
+    };
+
+    executor.exec(event_loop.run(event_handler).boxed());
+
+    Ok(actor)
+  }
+}
+
+impl Default for ActorBuilder {
+  fn default() -> Self {
+    Self::new()
+  }
+}
+
+/// Used to attach handlers and hooks to an [`ActorBuilder`].
+pub struct HandlerBuilder<'builder, MOD: SyncMode, OBJ>
+where
+  OBJ: Clone + Send + Sync + 'static,
+  MOD: 'static,
+{
+  pub(crate) object_id: ObjectId,
+  pub(crate) handler_map: &'builder mut HandlerMap,
+  _marker_obj: PhantomData<&'static OBJ>,
+  _marker_mod: PhantomData<&'static MOD>,
+}
+
+impl<'builder, OBJ> HandlerBuilder<'builder, Synchronous, OBJ>
+where
+  OBJ: Clone + Send + Sync + 'static,
+{
+  /// Add a synchronous handler function that operates on a shared state object and some
+  /// [`ActorRequest`]. The function will be called if the actor receives a request
+  /// on the given `endpoint` and can deserialize it into `REQ`. The handler is expected
+  /// to return an instance of `REQ::Response`.
+  pub fn add_sync_handler<REQ, FUT, FUN>(self, endpoint: &'static str, handler: FUN) -> Result<Self>
```

Comment:
Do you really need the generic `FUN` closure here or wouldn't it also work with a normal function pointer `fn(OBJ, Actor, RequestContext<REQ>) -> FUT`? None of your test cases ever capture anything, and therefore can be converted to function pointers.

Apart from that, you can also consider to get rid of `FUT` and instead do `fn(..) -> BoxedFuture<'static, REQ::Response>`, which itself directly implements `Send`, `Sync` and a `'static` lifetime.

Reply:
Thanks, I was able to get rid of all `FUN` generics. Very nice!

I kept `FUT` because requiring `BoxFuture` as a return type means users cannot simply provide a function pointer to some `async fn`, because such a function does not return `BoxFuture`. Instead, I think they'd have to implement a sync `fn` and return `BoxFuture` manually, which is too much of an ergonomic hit.